Source/Engine/Scripting/Coroutines/CoroutineSequence.h

```diff
@@ -0,0 +1,165 @@
+
+    /// <summary> Action to be executed. </summary>
+    API_EVENT()
+    Action OnRun;
+};
+
+/// <summary>
+/// Stores a predicate to be checked before the coroutine may continue execution.
+/// </summary>
+/// <remarks>
+/// This class is used due to limitations of Flax API.
+/// </remarks>
+API_CLASS(Sealed) class FLAXENGINE_API CoroutinePredicate final : public ScriptingObject
+{
+    DECLARE_SCRIPTING_TYPE_WITH_CONSTRUCTOR_IMPL(CoroutinePredicate, ScriptingObject);
+
+    /// <summary>
+    /// Predicate to be checked. Default value is <c>false</c>.
+    /// </summary>
+    /// <remarks>
+    /// It uses a reference to a boolean due to limitations of Flax API.
+    /// </remarks>
+    API_EVENT()
+    Delegate<bool&> OnCheck;
+};
+
+/// <summary>
+/// Utility class to store coroutine steps and build a coroutine.
+/// </summary>
+/// <remarks>
+/// Modifying the builder during the execution may cause undefined behavior.
+/// One builder may be shared between executions, reducing execution overhead.
+/// </remarks>
+API_CLASS(Sealed) class FLAXENGINE_API CoroutineSequence final : public ScriptingObject
+{
+    DECLARE_SCRIPTING_TYPE_WITH_CONSTRUCTOR_IMPL(CoroutineSequence, ScriptingObject);
+
+    /// <summary> Executes the code. </summary>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenRun(ScriptingObjectReference<CoroutineRunnable> runnable);
+
+    /// <summary>
+    /// Suspends the coroutine for the given number of seconds.
+    /// </summary>
+    /// <param name="seconds">
+    /// Minimal number of seconds to wait.
+    /// 0 is a valid value where the coroutine will be resumed on the closest time accumulation point.
+    /// </param>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitSeconds(float seconds);
+
+    /// <summary> Suspends the coroutine for the given number of frames. </summary>
+    /// <param name="frames">
+    /// Minimal number of <b>full</b> frames to wait.
+    /// 0 is a valid value where the coroutine will be resumed on the closest frame accumulation point.
+    /// </param>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitFrames(int32 frames);
+
+    /// <summary> Suspends the coroutine until the given suspend point. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenWaitForPoint(CoroutineSuspendPoint point);
+
+    /// <summary> Suspends the coroutine until the predicate is true. </summary>
+    API_FUNCTION()
+    ScriptingObjectReference<CoroutineSequence> ThenWaitUntil(ScriptingObjectReference<CoroutinePredicate> predicate);
+
+
+    /// <summary> QOL wrapper for running runnable action in native scripts. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenRunFunc(const Function<void()>& runnable);
+
+    /// <summary> QOL wrapper for waiting until predicate in native scripts. </summary>
+    ScriptingObjectReference<CoroutineSequence> ThenWaitUntilFunc(const Function<void(bool&)>& predicate);
```

Comment:
It can be `Function<bool()>` and `API_FUNCTION` so C# can pass `Action` directly.